ui/src/features/project/project-details/project-details.tsx

```diff
@@ -133,17 +131,69 @@ export const ProjectDetails = () => {
     };
     watchStages();
 
+    const watchWarehouses = async () => {
+      const promiseClient = createPromiseClient(KargoService, transport);
+      const stream = promiseClient.watchWarehouses({ project: name }, { signal: cancel.signal });
+      let warehouses = warehouseData?.warehouses || [];
+      const refresh = {} as { [key: string]: boolean };
+
+      for await (const e of stream) {
+        const index = warehouses.findIndex(
+          (item) => item.metadata?.name === e.warehouse?.metadata?.name
+        );
+        if (e.type === 'DELETED') {
+          if (index !== -1) {
+            warehouses = [
+              ...warehouses.slice(0, index),
+              ...warehouses.slice(index + 1)
+            ] as Warehouse[];
+          }
+        } else {
+          if (index === -1) {
+            warehouses = [...warehouses, e.warehouse as Warehouse] as Warehouse[];
+          } else {
+            warehouses = [
+              ...warehouses.slice(0, index),
+              e.warehouse as Warehouse,
+              ...warehouses.slice(index + 1)
+            ] as Warehouse[];
+          }
+        }
+
+        if (refresh[e.warehouse?.metadata?.name || '']) {
+          delete refresh[e.warehouse?.metadata?.name || ''];
+          refetchAvailableFreight();
+        }
+
+        const refreshing = e.warehouse?.metadata?.annotations['kargo.akuity.io/refresh'];
```

Comment:
If refresh annotation is found on a warehouse, we don't refetch freight yet. Instead, save it to a map which indicates it _started_ refreshing